_posts/2017-11-22-macros.md

```diff
@@ -0,0 +1,305 @@
+can come up with a robust solution.
+So far, we have seen two promising ideas to address or circumvent this problem,
+one by Martin Odersky and one by Lionel Parreaux.
+
+## Unification of LMS and Macros
+
+Martin Odersky recently shared [a gist](https://gist.github.com/odersky/f91362f6d9c58cc1db53f3f443311140)
+on Principled Metaprogramming for Scala.
+The design proposed in this gist is quite close to Meta OCaml as well as to a
+calculus by Davies and Pfenning inspired by modal logic, see
+["A modal analysis of staged computation"](https://dl.acm.org/citation.cfm?doid=382780.382785).
+
+This proposal has triggered a rich discussion in the gist, which you may find interesting.
+However, the proposal may go too far from current scala-reflect macros to
+cause a migration concern.
+Quoting Martin Odersky from
+[r/scala](https://www.reddit.com/r/scala/comments/7emo4s/oderskys_ideas_about_a_new_design_for_macros_that/dq6m1da/)
+
+> One attractive aspect of the system is that it works equally well for staging
+> and for macros.  But as a macro system it is quite restrictive because it
+> imposes a strong typing discipline, only works for full expressions, and in
+> its original version does not allow for code inspection. Lionel's Parreaux's
+> Squid system (paper to appear at upcoming POPL) is more powerful in that it
+> does allow inspection.
+
+Lionel Parreaux ([@LPTK]) is a PhD student at EPFL who has been working on
+related problems for the past couple years as part of his research on the
+intersection of database systems, programming languages and large-scala data
+analysis.
+
+## Squid: reusable and type-safe quasiquotes
+
+[Squid] is a metaprogramming framework that facilitates the type-safe
+manipulation of Scala programs through quasiquotes.
+Unlike scala-reflect quasiquotes, Squid quasiquotes are statically guaranteed
+to be well-typed, well-scoped and hygienic.
+The properties of Squid quasiquotes make them ideal for robust and safe
+transformation of trees, the exact problem we struggled with in our prototype
+macro system.
+
+One important attribute of Squid quasiquotes is that they are reusable.
+Similar to scala-reflect quasiquotes, it's possible to write generic utility
+functions with Squid quasiquotes and compose them into more complex tree
+structures.
+Many alternative statically typed quasiquote systems do not support this
+ability to combine small quasiquote fragments into larger programs.
+This functionality is enabled by Squid's first-class encoding of free
+variables.
+
+Squid quasiquotes statically track the context/scope of a quasiquote in its type to
+prevent references to unbound identifiers.
+For example, `code"(x: Int) + 1"` would be rejected at compile-time since `x`
+is undefined.
+Instead, quasiquotes must explicitly annotate free variables
+using question marks `code"(?x: Int) + 1)`.
+Free variables are then tracked in the second parameter of quasiquote's type
+`Code[Int, {x: Int}]`.
```

Comment:
Note that last I checked, this is the way Dotty print structural types, but the user can't write it like that –– it needs a `val` as in `{val x: Int}`. Not sure which one should be shown here. What do you think?

Reply:
Added `val` :)